test/metabase/query_processor_test/date_time_zone_functions_test.clj

```diff
@@ -319,3 +319,206 @@
                           :fields [[:field (mt/id :times :index)]]}}]]
         (testing title
           (is (= (set expected) (set (test-datetime-math query)))))))))
+
+(mt/defdataset temporal-dataset
+  [["datediff-mixed-types"
+    [{:field-name "index" :base-type :type/Integer}
+     {:field-name "description" :base-type :type/Text}
+     {:field-name "tz" :base-type :type/DateTimeWithTZ}
+     {:field-name "d" :base-type :type/Date}
+     {:field-name "dt" :base-type :type/DateTime}]
+    [[1 "simple comparing across types" #t "2021-08-03T08:09:10.582Z" #t "2022-09-04" #t "2022-10-05 09:19:09"]]]])
```

Comment:
If you just want to compare different temporal types like `timestamp` vs `datetime` vs `date` vs `timestamptz`, we have the existing `attempted-murders` dataset that has basically one of everything in it -- you can use that instead of creating a one-off dataset here

Reply:
Updated in https://github.com/metabase/metabase/commit/1b9f0ab9eb3bfdded2ae976fa7ff6e64fa548791